plugins/inputs/dpdk/README.md

```diff
@@ -0,0 +1,184 @@
+This configuration allows getting metrics from two separate applications exposing their telemetry interfaces
+via separate sockets. For each plugin instance a unique tag `[inputs.dpdk.tags]` allows distinguishing between them. 
+```toml
+# Instance #1 - L3 Forwarding with Power Management Application
+[[inputs.dpdk]]
+  socket_path = "/var/run/dpdk/rte/l3fwd-power_telemetry.v2"
+  device_types = ["ethdev"]
+  additional_commands = ["/l3fwd-power/stats"]
+
+  [inputs.dpdk.ethdev]
+    exclude_commands = ["/ethdev/link_status"]
+
+  [inputs.dpdk.tags]
+    dpdk_instance = "l3fwd-power"
+
+# Instance #2 - L2 Forwarding with Intel Cache Allocation Technology (CAT) Application
+[[inputs.dpdk]]
+  socket_path = "/var/run/dpdk/rte/l2fwd-cat_telemetry.v2"
+  device_types = ["ethdev"]
+
+[inputs.dpdk.ethdev]
+  exclude_commands = ["/ethdev/link_status"]
+
+  [inputs.dpdk.tags]
+    dpdk_instance = "l2fwd-cat"
+```
+This utilizes Telegraf's standard capability of [adding custom tags](https://github.com/influxdata/telegraf/blob/master/docs/CONFIGURATION.md#input-plugins)
+to input plugin's measurements.
+
+## Metrics
+The DPDK socket accepts `command,params` requests and returns metric data in JSON format. All metrics from DPDK socket
+become flattened using [Telegraf's JSON Flattener](../../parsers/json/README.md) and exposed as fields. 
+If DPDK response contains no information (is empty or is null) then such response will be discarded.
+
+> **NOTE:**  Since DPDK allows registering custom metrics in its telemetry framework the JSON response from DPDK 
+> may contain various sets of metrics. While metrics from `/ethdev/stats` should be most stable, the `/ethdev/xstats`
+> may contain driver-specific metrics (depending on DPDK application configuration). The application-specific commands
+> like `/l3fwd-power/stats` can return their own specific set of metrics.  
+
+## Example output
+The output consists of plugin name (`dpdk`), and a set of tags that identify querying hierarchy:
+```
+dpdk,host=dpdk-host,dpdk_instance=l3fwd-power,command=/ethdev/stats,params=0 [fields] [timestamp]
+```
+
+| Tag | Description |
+|-----|-------------|
+| `host` | hostname of the machine (consult [Telegraf Agent configuration](https://github.com/influxdata/telegraf/blob/master/docs/CONFIGURATION.md#agent) for additional details) |
+| `dpdk_instance` | custom tag from `[inputs.dpdk.tags]` (optional) | 
+| `command` | executed command (without params) |
+| `params` | command parameter, e.g. for `/ethdev/stats` it is the id of NIC as exposed by `/ethdev/list`<br>For DPDK app that uses 2 NICs the metrics will output e.g. `params=0`, `params=1`. |
+
+When running plugin configuration below...
+```toml
+[[inputs.dpdk]]
+  device_types = ["ethdev"]
+  additional_commands = ["/l3fwd-power/stats"]
+  [inputs.dpdk.tags]
+    dpdk_instance = "l3fwd-power"
+```
+
+...expected output for `dpdk` plugin instance running on host named `host=dpdk-host`:
```

Comment:
The observation is correct, list of fields changes depending on the command being run and also the amount and the naming of metrics vary across different drivers and apps. This is yet another effect of the freedom given by DPDK to the libraries and applications to expose their specific telemetry. Such custom metrics can be exposed from libraries e.g. through `xstats` ("extended statistics" that are by nature customizable) or through [registering its own telemetry commands](https://doc.dpdk.org/guides-20.05/prog_guide/telemetry_lib.html#registering-commands) which makes the set of exposed metrics highly dependent on the driver that is used and on the DPDK-based application. So right now this leaves it to the drivers and DPDK applications to maintain metrics compatibility across versions.

E.g. looking at [feature availability in networking drivers for DPDK](https://doc.dpdk.org/guides/nics/overview.html) we can tell how much differences there are between drivers and such differences are also observed in set of exposed metrics. There are no guarantees that metrics will be exposed in the same way across all the drivers or applications, even if they effectively intend to expose the same piece of information. 

There are ongoing efforts to standardize at least some of the common metrics (e.g. `/ethdev/stats` introduced in DPDK 20.11) but even the DPDK telemetry interface itself is still considered "experimental" with a note that "the output format may change in the future", as stated in [DPDK Telemetry User Guide](https://doc.dpdk.org/guides-20.05/howto/telemetry.html).

So since there's not many constants to currently rely on, we went with generic approach of reading metrics in this plugin and just expose what DPDK-based applications and underlying drivers are exposing.

Reply:
No changes here, as we agreed.